kernel/kernel-api/src/main/java/io/delta/kernel/internal/icebergcompat/IcebergCompatMetadataValidatorAndUpdater.java

```diff
@@ -151,10 +166,127 @@ Optional<Metadata> validateAndUpdate(
    * Defines checks for compatibility with the targeted iceberg features (icebergCompatV1 or
    * icebergCompatV2 etc.)
    */
-  interface IcebergCompatCheck {
+  protected interface IcebergCompatCheck {
     void check(IcebergCompatInputContext inputContext);
   }
 
+  ///////////////////////////////////////////////////////////
+  /// Implementation of {@link IcebergCompatCheck}        ///
+  ///////////////////////////////////////////////////////////
+  protected static final IcebergCompatCheck CHECK_NO_COMPAT_V1_ENABLED =
+      (inputContext) -> {
+        if (Boolean.valueOf(
+            inputContext
+                .newMetadata
+                .getConfiguration()
+                .getOrDefault("delta.enableIcebergCompatV1", "false"))) {
+          throw DeltaErrors.icebergCompatIncompatibleVersionEnabled(
+              inputContext.compatFeatureName, "delta.enableIcebergCompatV1");
+        }
+      };
+
+  protected static final IcebergCompatCheck V2_CHECK_HAS_SUPPORTED_TYPES =
+      (inputContext) -> {
+        List<Tuple2<List<String>, StructField>> matches =
+            SchemaUtils.filterRecursively(
+                inputContext.newMetadata.getSchema(),
+                /* recurseIntoMapAndArrayTypes= */ true,
+                /* stopOnFirstMatch = */ false,
+                field -> {
+                  DataType dataType = field.getDataType();
+                  return !(dataType instanceof ByteType
+                      || dataType instanceof ShortType
+                      || dataType instanceof IntegerType
+                      || dataType instanceof LongType
+                      || dataType instanceof FloatType
+                      || dataType instanceof DoubleType
+                      || dataType instanceof DecimalType
+                      || dataType instanceof StringType
+                      || dataType instanceof BinaryType
+                      || dataType instanceof BooleanType
+                      || dataType instanceof DateType
+                      || dataType instanceof TimestampType
+                      || dataType instanceof TimestampNTZType
+                      || dataType instanceof ArrayType
+                      || dataType instanceof MapType
+                      || dataType instanceof StructType);
+                });
+
+        if (!matches.isEmpty()) {
+          throw DeltaErrors.icebergCompatUnsupportedTypeColumns(
+              inputContext.compatFeatureName,
+              matches.stream().map(tuple -> tuple._2.getDataType()).collect(toList()));
+        }
+      };
+
+  protected static final IcebergCompatCheck CHECK_HAS_ALLOWED_PARTITION_TYPES =
```

Comment:
add a comment, these are the common supported partition types for compatv2 and v3